modules/quality/include/opencv2/quality/QualityGMSD.hpp

```diff
@@ -0,0 +1,89 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#ifndef OPENCV_QUALITY_QUALITYGMSD_HPP
+#define OPENCV_QUALITY_QUALITYGMSD_HPP
+
+#include "QualityBase.hpp"
+
+namespace cv {
+    namespace quality {
+
+        namespace detail {
+
+            namespace gmsd {
+                using mat_type = UMat;
+
+                // holds computed values for an input mat
+                struct mat_data
+                {
+                    mat_type
+                        gradient_map
+                        , gradient_map_squared
+                        ;
+
+                    mat_data(const mat_type&);
+                };  // mat_data
+            }   // gmsd
+        }   // detail
+
+        /**
+        @brief Full reference GMSD algorithm
+        http://www4.comp.polyu.edu.hk/~cslzhang/IQA/GMSD/GMSD.htm
+        */
+        class CV_EXPORTS_W QualityGMSD
+            : public QualityBase {
+        public:
+
+            /**
+            @brief Compute GMSD
+            @param cmpImgs Comparison images
+            @returns Per-channel GMSD
+            */
+            CV_WRAP cv::Scalar compute(InputArrayOfArrays cmpImgs) CV_OVERRIDE;
+
+            /** @brief Implements Algorithm::empty()  */
+            CV_WRAP bool empty() const CV_OVERRIDE { return _qualityMaps.empty() && _refImgData.empty(); }
+
+            /** @brief Implements Algorithm::clear()  */
+            CV_WRAP void clear() CV_OVERRIDE { _qualityMaps.clear(); _refImgData.clear(); Algorithm::clear(); }
+
+            /** @brief Returns pointer to output quality maps images that were generated during computation, if supported by the algorithm.  */
+            CV_WRAP const std::vector<quality_map_type>& getQualityMaps() const CV_OVERRIDE { return _qualityMaps; }
```

Comment:
OpenCV approach is:
```
CV_WRAP void getQualityMaps(OutputArrayOfArrays dst);
```

In the current form it is not handled properly by Python/Java bindings generator.